src/useQuery.ts

```diff
@@ -51,101 +49,115 @@ export interface QueryHookResult<TData, TVariables>
 
 export function useQuery<TData = any, TVariables = OperationVariables>(
   query: DocumentNode,
-  { suspend = true, ...restOptions }: QueryHookOptions<TVariables> = {}
+  {
+    // Hook options
+    suspend = true,
+
+    // Watch options
+    pollInterval,
+    notifyOnNetworkStatusChange,
+
+    // Apollo client options
+    context,
+    metadata,
+    variables,
+    fetchPolicy,
+    errorPolicy,
+    fetchResults,
+  }: QueryHookOptions<TVariables> = {}
 ): QueryHookResult<TData, TVariables> {
   const client = useApolloClient();
-  const [result, setResult] = useState<null | QueryHookState<TData>>(null);
-  const previousQuery = useRef<null | DocumentNode>(null);
-  const previousRestOptions = useRef<null | QueryHookOptions<TVariables>>(null);
-  const observableQuery = useRef<null | ObservableQuery<TData, TVariables>>(
-    null
+
+  const watchQueryOptions: WatchQueryOptions<TVariables> = useMemo(
+    () => ({
+      context,
+      errorPolicy,
+      fetchPolicy,
+      fetchResults,
+      metadata,
+      notifyOnNetworkStatusChange,
+      pollInterval,
+      query,
+      variables,
+    }),
+    [
+      query,
+
+      pollInterval,
+      notifyOnNetworkStatusChange,
+
+      context,
```

Comment:
Both `context` and `metadata` can be objects, so we should use `objToKey` also here. I think we should also change `objToKey` to correctly support all falsy values:

```typescript
export function objToKey(obj: Record<string, any>) {
  if (!obj) {
    return obj;
  }
  // ...
}

Reply:
I think we should encourage people to memoize these values by themselves, it's really easy to do with hooks: 
```js
const { accessToken } = useAuthPair();
const queryContext = useMemo(
  () => ({
    headers: {
      authorization: !accessToken ? undefined : `Bearer ${accessToken}`
    }
  }),
  [accessToken]
);
```

Also user can pass ANYTHING to this objects like regex or functions so `objToKey` would create false keys.

It's better to extend documentation about these properties.

Reply:
I think we should provide good defaults for most cases. With the implementation from this PR is easy to miss that and in that case, no memoization will be used. In the current implementation on master [both `context` and `metadata` are serialized](https://github.com/trojanowski/react-apollo-hooks/blob/a303a6058ef1800ef360521f5798af3bfdef9e10/src/useQuery.ts#L75) (as a part of `restOptions`) and they still would be, even after merging this PR [in `queryCache`](https://github.com/trojanowski/react-apollo-hooks/blob/a303a6058ef1800ef360521f5798af3bfdef9e10/src/queryCache.ts#L50). We haven't had problems with that so far. We could tweak `objToKey` a little to better support non-objects (and also rename it) e. g.:

```typescript
export function optionsToKey(optionOrOptions: any): any {
  if (!isPlainObject(optionOrOptions)) {
    return optionOrOptions;
  }
  const sortedObj = Object.keys(optionOrOptions)
    .sort()
    .reduce((result: Record<string, any>, key) => {
      result[key] = optionsToKey(optionOrOptions[key]);
      return result;
    }, {});
  return JSON.stringify(sortedObj);
}
```